Python/immutability.c

```diff
@@ -156,281 +239,877 @@ static bool is_c_wrapper(PyObject* obj){
+    // debug("----\n");
+#else
+    (void)state;
+#endif
+}
 
-    if (push(state->visited_list, op) != 0)
-    {
-        // If we fail to add the item to the visited set, then we
-        // will not be able to backtrack, so go to error case.
-        goto error;
-    }
+// During the freeze, we removed the reference counts associated
+// with the internal edges of the SCC.  This visitor detects these
+// internal edges and re-adds the reference counts to the
+// objects in the SCC.
+static int scc_add_internal_refcount_visit(PyObject* obj, void* curr_root)
+{
+    if (obj == NULL)
+        return 0;
+
+    // Ignore mutable outgoing edges.
+    if (!_Py_IsImmutable(obj))
+        return 0;
+
+    // Find the scc root.
+    PyObject* root = scc_root(obj);
+
+    // If it is different SCC, then we can ignore it.
+    if (root != curr_root)
+        return 0;
+
+    // Increase the reference count as we found an interior edge for the SCC.
+    debug_obj("Reinstate %s (%p) with rc %zu from %p\n", obj, Py_REFCNT(obj), curr_root);
+    obj->ob_refcnt++;
 
-    _Py_SetImmutable(op);
     return 0;
+}
 
-error:
-    PyErr_SetString(PyExc_RuntimeError, "Failed to add item to visited set");
-    return -1;
+struct SCCDetails {
+    int has_weakreferences;
+    int has_legacy_finalizers;
+    int has_finalizers;
+};
+
+static void scc_set_refcounts_to_one(PyObject* obj)
+{
+    PyObject* n = obj;
+    do {
+        PyObject* c = n;
+        n = scc_next(c);
+        c->ob_refcnt = 1;
+    } while (n != obj);
+}
+
+static void scc_reset_root_refcount(PyObject* obj)
+{
+    assert(scc_root(obj) == obj);
+    size_t scc_rc = _Py_REFCNT(obj) * 2;
+    PyObject* n = obj;
+    do {
+        PyObject* c = n;
+        n = scc_next(c);
+        scc_rc -= _Py_REFCNT(c);
+    } while (n != obj);
+    obj->ob_refcnt = scc_rc;
 }
 
-// Called on the failure of a freeze operation.
-// This unsets the immutability of all the objects that were visited.
-void fail_freeze(struct FreezeState *state)
+// This will restore the reference counts for the interior edges of the SCC.
+// It calculates some properites of the SCC, to decide how it might be
+// finalised.  Adds an RC to every element in the SCC.
+static void scc_add_internal_refcounts(PyObject* obj, struct SCCDetails* details)
 {
-    Py_XDECREF(state->dfs);
+    assert(_Py_IsImmutable(obj));
+    PyObject* root = scc_root(obj);
+
+    details->has_weakreferences = 0;
+    details->has_legacy_finalizers = 0;
+    details->has_finalizers = 0;
+
+    // Add back the reference counts for the interior edges.
+    PyObject* n = obj;
+    do {
+        debug_obj("Unfreezing %s @ %p\n", n);
+        PyObject* c = n;
+        n = scc_next(c);
+        //  WARNING
+        //  CHANGES HERE NEED TO BE REFLECTED IN freeze_visit
+
+        if (PyType_Check(c)) {
+            // TODO(Immutable): mjp: Special case for types not sure if required. We should review.
```

Comment:
Likely required until we have a custom `tp_reachable` function for Pyrona.

This will likely also cover parts of the special case for weak references, making all of this code way cleaner :D